### PR TITLE
SSE: Localize/Contain Errors within an Expression

### LINKED ISSUE #71772: SSE: Contain Errors so one error doesn't make everything error
Problem:

If any data source query or expression returns an error, all expressions and datasource queries will fail.

This is annoying when trying to write expressions, and also makes it less useful with dashboards as things become more fragile.

How to improve:
- For each "group", will only the fail the group. Where a "group" is a single graph as parsed by server side expressions. Or to put it another way - any stand alone expression or data source query. Or an expression and its dependencies.
- Fancier would be within a group, we walk the graph as far as we can until blocked by errors

I'm not quite sure how difficult this will be. I think the main things:
 - Make sure things don't execute if there inputs have errored, and return some sort of dependency error
 - Making this a return type. As I expect this will lead to a feature request of being able to replace errors with some sort of value.

### PR DESCRIPTION
**Pull request by kylebrandt**
**What is this feature?**
Currently if anything single operation with an Expression fails, the whole thing fails.

This changes it so only operations, or operations that depend on them, will return an error.

image

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/71772

**Special notes for your reviewer:**

- [ ] Need alerting FE Grafana Manager Alert edit to show errors associated with each query (See Screenshot below)
- [ ] See if works with Recorded Queries
- [ ] Make sure instrumentation and other things that were (or are still) using `defer func()` are not messed up and capturing things wrong.

**Alerting UI Screenshot**
image

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/expr/graph.go [hunk 1]
```diff
@@ -55,7 +55,45 @@ type DataPipeline []Node
 // map of the refId of the of each command
 func (dp *DataPipeline) execute(c context.Context, now time.Time, s *Service) (mathexp.Vars, error) {
 	vars := make(mathexp.Vars)
+
+	// Execute datasource nodes first, and grouped by datasource.
```
Comment (kylebrandt) [inline pkg/expr/graph.go:59]: The parts related to doing the datasource first come from another PR that this one based of off, mentioned in the PR original issue notes.

File: pkg/expr/graph.go [hunk 2]
```diff
@@ -55,7 +55,45 @@ type DataPipeline []Node
...
+	for _, node := range *dp {
+		if node.NodeType() != TypeDatasourceNode {
+			continue
+		}
+		dsNodes = append(dsNodes, node.(*DSNode))
+	}
+
+	executeDSNodesGrouped(c, now, vars, s, dsNodes)
+
+	// TODO: Mark Dependent nodes of vars[X] that have result.Error != nil, and then don't execute dependent nodes,
```
Comment (kylebrandt) [inline pkg/expr/graph.go:70]: rm comment

File: pkg/expr/graph.go [hunk 3]
```diff
@@ -55,7 +55,45 @@ type DataPipeline []Node
...
+			continue // already executed via executeDSNodesGrouped
+		}
+
+		var hasDepError bool
+		if cmd, ok := node.(*CMDNode); ok {
+			for _, neededVar := range cmd.Command.NeedsVars() {
+				if res, ok := vars[neededVar]; ok {
+					if res.Error != nil {
+						errResult := mathexp.Results{
+							Error: fmt.Errorf("not executing expression %v because expression %v errored", node.RefID(), neededVar), // TODO errutil
```
Comment (yuri-tceretian) [inline pkg/expr/graph.go:84]: It can be a public variable so in alerting we are able to filter them out and report only the root cause.

File: pkg/expr/service.go [hunk 4]
```diff
@@ -109,6 +109,7 @@ func (s *Service) ExecutePipeline(ctx context.Context, now time.Time, pipeline D
 	for refID, val := range vars {
 		res.Responses[refID] = backend.DataResponse{
 			Frames: val.Values.AsDataFrames(refID),
+			Error: val.Error,
```
Comment (kylebrandt) [inline pkg/expr/service.go:112]: This is where the main impact on alerting is:

```
vars, err := pipeline.execute(ctx, now, s)
	if err != nil {
		return nil, err
	}
	for refID, val := range vars {
		res.Responses[refID] = backend.DataResponse{
			Frames: val.Values.AsDataFrames(refID),
			Error: val.Error,
```

This is because where before we would return an error from the service (with pipeline execution), we generally are not doing that anymore. Instead, the error moves to the property of each Response, and then gets passed to the caller (alerting).

So alerting needs to check for this condition on evaluation. Also, for the `eval` endpoint, the UI needs to show these errors.

So in short, unless alerting starts checking for the Error on Response(s) - in particular the condition in the case of evaluation, and all expressions for the UI, what was previously shown as errors will become masked.

File: pkg/expr/graph.go (hunk 3 above)
Comment (kylebrandt) [reply pkg/expr/graph.go]: I adjusted eval.go in alerting to walk the Dependency Errors and return the underlying non-dependency error.

**Review by yuri-tceretian — COMMENTED** [review]
The backend works as expected. However, I have two comments:
1. I think we can refactor the code a bit to make it more straightforward and delegate some responsibilities to nodes.
2. I could not find tests for the new logic. Can you please add tests or point to existing ones.

UI side I think needs a little bit work:
1. On the panel config I noticed that it does not show error for each query 
<details><summary> Example </summary>

image
image

</details>

2. I like the badges we show for each query but I worry if they are visible enough. It would be great to make the button `Preview` a bit more interactive: when user clicks on it it gets disabled and a shows a spinner. When the eval result is received it could have some animation and blink with red or green depends on the outcome. Alternative, is to display toast with the first error in the response or all errors that are related to data source nodes and then fallback to the first one that belongs to an expression.

<details><summary>Current PR behavior</summary>

https://github.com/grafana/grafana/assets/25988953/f58c1b3d-6eb9-41a6-90b4-0a88cd11524a

</details>

3. Showing errors for dependent nodes is a little bit frightening. Inexperienced user could not notice that the error is actually from the depenent node. It requires to hover on error to figure that out. I wonder if we can display them differently, for example, as a warning.

File: pkg/expr/graph.go [hunk 5]
```diff
@@ -67,15 +72,34 @@ func (dp *DataPipeline) execute(c context.Context, now time.Time, s *Service) (m
...
 	}
 
 	for _, node := range *dp {
 		if groupByDSFlag && node.NodeType() == TypeDatasourceNode {
 			continue // already executed via executeDSNodesGrouped
 		}
+
+		// Don't execute nodes that have dependent nodes that have failed
+		var hasDepError bool
+		if cmd, ok := node.(*CMDNode); ok {
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply pkg/expr/graph.go:85]: > Can you please explain why we need to assert type here? Afaiu, as long as NeedsVars() returns some result, it is considered to be dependent, regardless its type.

Because `NeedsVars` is not on the Node interface, but will look at moving it there as that might make things cleaner (Perhaps not in this PR, will see...)

> Also, I wonder if we can move this into nodes, i.e. when we call Execute of a node that has dependency, the node itself checks that all related nodes' results are successful and return DependencyError if they are not.

I like it at this level _for now_ for the simplicity of it being in one place and not having to call it. That being said, I do expect this will change to be the way you describe it down the road. This is because I imagine there may be requests to do something to handle errors (like with math's nodata funcs for nodata).

[the remark this answers is not in the document]
Comment (kylebrandt) [reply pkg/expr/graph.go]: Added NeedsVars

**Review by yuri-tceretian — APPROVED** [review]
LGTM. 

Kyle and I had a chat and agreed to merge it as it is right now and not refactor it further because otherwise, the PR would grow big. Later we will refactor this code.
